### ISSUE TITLE
Python 3.13 support TODOs

### ISSUE THREAD

**Issue by nitzmahone** [python-cffi/cffi#77]
Now that Python 3.13 has hit beta, we're overdue for at least a pre-release that supports it. Prerequisite to getting that out the door:

- [x] Implement dynamic GHA matrix to use a single workflow under the primary repo/org for all CI and release activity.
- [x] Implement support for GHA-hosted Apple Silicon runners.
- [x] Add aarch64 musllinux wheel builds.
- [ ] Fix [broken static linking of libffi](https://github.com/nitzmahone/cffi/actions/runs/9126860324/job/25095961084) in `musllinux` builds.
- [ ] Determine how to handle [test failures from null pointer subtraction errors](https://github.com/nitzmahone/cffi/actions/runs/9126860324/job/25095973663) in newer versions of clang (ignore those tests, disable the error, change the code?)

**Comment by arigo** [comment]
re [test failures from null pointer subtraction errors](https://github.com/nitzmahone/cffi/actions/runs/9126860324/job/25095973663): I've got a quick hack that seems to work fine with MSVC, can you check if it does the trick on clang?  The hack is:

```diff --git a/src/cffi/recompiler.py b/src/cffi/recompiler.py
index 4167bc05..ac6c163e 100644
--- a/src/cffi/recompiler.py
+++ b/src/cffi/recompiler.py
@@ -953,7 +953,7 @@ class Recompiler:
                 if cname is None or fbitsize >= 0:
                     offset = '(size_t)-1'
                 elif named_ptr is not None:
-                    offset = '((char *)&((%s)0)->%s) - (char *)0' % (
+                    offset = '((char *)&((%s)4096)->%s) - (char *)4096' % (
                         named_ptr.name, fldname)
                 else:
                     offset = 'offsetof(%s, %s)' % (tp.get_c_name(''), fldname)
```

It's not handling null pointers any more...  and this expression is only used in a struct initialization, not in real code.

**Comment by alan-isaac** [comment]
Apologies if this proves irrelevant; I'm just a user. Currently trying to pip install (Python 3.13.0b1, Win 11) fails with errors that include:
``_cffi_backend.obj : error LNK2001: unresolved external symbol _PyErr_WriteUnraisableMsg``

**Comment by nitzmahone** [comment]
@alan-isaac Yeah, that's expected from released bits- assuming your box has the right tools already installed, you should be able to do `pip install https://github.com/python-cffi/cffi/archive/refs/heads/main.tar.gz` to build the extension from the 3.13-compatible bits we're preparing to release.

**Comment by nitzmahone** [comment]
@arigo yep, changing the named pointer stuff in recompiler [as you suggested](https://github.com/python-cffi/cffi/issues/77#issuecomment-2122161275) to use matching non-zero offsets prevents the null pointer subtraction warnings/errors on the [newer clang releases](https://github.com/nitzmahone/cffi/actions/runs/9212858811/job/25345474070#step:5:185).

I've filed that as PR #78, since it's not just a test/build change.

Thanks!

**Comment by ofek** [comment]
For the static linking release criteria, this issue I opened a few weeks ago might provide useful context https://github.com/python-cffi/cffi/issues/75

**Comment by jaraco** [comment]
Are there plans to cut a release with these fixes? My system (Python 3.13b2 on macOS ARM) fails to install cffi. Building against github works around the problem.
